apis/python/src/tiledbsoma/io/registration/signatures.py

```diff
@@ -0,0 +1,308 @@
+            if "varm" in exp.ms[measurement_name]:
+                for varm_layer_name in exp.ms[measurement_name].varm.keys():
+                    varm = exp.ms[measurement_name].varm[varm_layer_name]
+                    varm_dtypes[varm_layer_name] = str(
+                        varm.schema.field("soma_data").type
+                    )
+
+            return cls(
+                obs_schema=obs_schema,
+                var_schema=var_schema,
+                X_dtypes=X_dtypes,
+                raw_X_dtype=raw_X_dtype,
+                raw_var_schema=raw_var_schema,
+                obsm_dtypes=obsm_dtypes,
+                varm_dtypes=varm_dtypes,
+            )
+
+    @classmethod
+    def check_compatible(cls, signatures: Dict[str, Self]) -> None:
+        """
+        Determines if any number of signatures from SOMA experiment or AnnData/H5AD will be safe
+        from schema-incompatibility at ingestion time. On failure, a ``ValueError`` is raised
+        with user-suitable cause details.
+        """
+        if len(signatures) < 2:
+            return
+        items = list(signatures.items())
+        namea, siga = items[0]
+        for nameb, sigb in items[1:]:
+            if not siga._compatible_with(sigb):
+                raise ValueError(
+                    f"Incompatible signatures {namea!r}, {nameb!r}:\n{siga.toJSON()}\n{sigb.toJSON()}"
+                )
+
+    def _compatible_with(self, other: Self) -> bool:
+        """
+        Pairwise helper method for ``compatible``. Reasons for incompatibility are currently advised
+        to be handled a level up by simply showing the user the failed signature pair.
+        """
+
+        # Implementation note: _at present_ this could be implemented as `self == other`.  But
+        # "coming soon" we'll allow people the ability to do things like coercing one input's
+        # float64 to another's float32 and the evolution will be toward more iffing, not less.  As
+        # well, in that case, we'll need to also fine-grain the error-reporting to clearly spell out
+        # what we coudn't handle -- rather than just showing the user the two signatures' .toJSON()
+        # output as we do today.
+
+        if self.obs_schema != other.obs_schema:
+            return False
+
+        if self.var_schema != other.var_schema:
+            return False
+
+        if self.X_dtypes != other.X_dtypes:
+            return False
+
+        if self.raw_X_dtype != other.raw_X_dtype:
+            return False
+        if self.raw_var_schema != other.raw_var_schema:
+            return False
+
+        if self.obsm_dtypes != other.obsm_dtypes:
+            return False
+        if self.varm_dtypes != other.varm_dtypes:
+            return False
+
+        return True
+
+    def toJSON(self) -> str:
+        """Presents a signature as JSON which is suitable for distributed logging."""
+        return json.dumps(
+            self, default=lambda o: attrs.asdict(o), sort_keys=True, indent=4
```

Comment:
This seems like it could be `json.dumps(attrs.asdict(self), ...)` rather than setting `default=`. Am I missing something?

(If there is a reason to set `default`, then it can be shortened to `default=attrs.asdict`, since that’s all the lambda does.)

Reply:
It is necessary to exist; I get errors without. I'll simplify the syntax as you suggest.